Sources/Utility/Process.swift

```diff
@@ -0,0 +1,253 @@
+            posix_spawn_file_actions_adddup2(&fileActions, 2, 2)
+        }
+
+        let argv: [UnsafeMutablePointer<CChar>?] = arguments.map{ $0.withCString(strdup) }
+        defer { for case let arg? in argv { free(arg) } }
+        
+        let env: [UnsafeMutablePointer<CChar>?] = environment.map{ "\($0.0)=\($0.1)".withCString(strdup) }
+        defer { for case let arg? in env { free(arg) } }
+
+        rv = posix_spawnp(&processID, argv[0], &fileActions, &attributes, argv + [nil], env + [nil])
+
+        posix_spawn_file_actions_destroy(&fileActions)
+        posix_spawnattr_destroy(&attributes)
+
+        guard rv == 0 else {
+            print("spawn error")
+            throw Error.error
+        }
+    }
+
+    /// Blocks the calling process until the subprocess finishes execution.
+    ///
+    /// - Throws: Process.Error
+    public func waitUntilExit() throws {
+        assert(launched, "The process is not yet launched.")
+        // FIXME: We should probably read the output from somewhere else, maybe in a seperate thread?
+        if saveOutput {
+            try readOutput()
+        }
+        var exitStatus: Int32 = 0
+        var result = waitpid(processID, &exitStatus, 0)
+        while (result == -1 && errno == EINTR) {
+            result = waitpid(processID, &exitStatus, 0)
+        }
+        if result == -1 {
+            throw Error.error
+        }
+    }
+
+    /// Reads the output from save output pipes.
+    ///
+    /// - Throws: Process.Error
+    private func readOutput() throws {
+        assert(saveOutput, "Save output is false.")
+        assert(saveOutputPipe != [0, 0], "Pipes not initialized.")
+        // Close the write end of the output pipe.
+        guard close(saveOutputPipe[1]) == 0 else {
+            print("close pipe error")
+            throw Error.error
+        }
+
+        // Read all of the data from the output pipe.
+        let N = 4096
+        var buf = [Int8](repeating: 0, count: N + 1)
+        
+        var output = ""
+        loop: while true {
+            let n = read(saveOutputPipe[0], &buf, N)
+            switch n {
+            case  -1:
+                if errno == EINTR {
+                    continue  // try again!
+                } else {
+                    print("read output err1")
+                    throw Error.error
+                }
+            case 0:
+                break loop
+            default:
+                buf[n] = 0 // must null terminate
+                if let str = String(validatingUTF8: buf) {
+                    output += str
+                } else {
+                    print("read output err2")
+                    throw Error.error
+                }
+            }
+        }
+        self.output = output
+        // Close the read end of the output pipe.
+        close(saveOutputPipe[0])
+    }
+
+    public func interrupt() {
+        assert(launched, "The process is not yet launched.")
+        _ = libc.kill(processID, SIGINT)
```

Comment:
This doesn't send the signal to the whole process group. We need to either use `-processID` or `killpg` explicitly.